### ISSUE TITLE
Post/Pre processing and load hooks

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#724]
We need a more explicit way for a pre/post preparation or load hook instead of overloading methods of GraphHopper - or is this okay and needs just some more tests and docs? 

Should we just use a listener list instead?
```java
graphHopper.addHook(new Hook() {
        boolean beforeLoad(/* what args? */) { /*return false if load should not be called*/ }
        boolean beforeImport(/* what args? */) {}
        boolean beforePreprocessing(/* what args? */) {}
        void afterLoad(/* what args? */) {}
        void afterImport(/* what args? */) {}
        void afterPreparation(/* what args? */) {}
});
```

See #700 and https://github.com/karussell/graphhopper-osm-id-mapping/issues/1 for the issue behind this.

**Comment by devemux86** [comment]
Will that come with removal of overloading options (i.e. protected / public methods) ?
Because they can be useful in many situations where we need customization options.

Listener approach is an interesting one. Another event could be about the arguments initialization.

**Comment by karussell** [comment]
> Will that come with removal of overloading options (i.e. protected / public methods) ?

Yes, I was hoping to make the methods final then - at least for the ones where we have a hook.

> Because they can be useful in many situations where we need customization options.

Which options are these?

> Another event could be about the arguments initialization.

Do you mean something like afterInit? Isn't this identical to e.g. `beforeLoad`?

**Comment by devemux86** [comment]
>Which situations do you mean here?

I have no actual cases here, like that which initiated this.
Just with removing functionality worries me when we may need it in the future. :smile:

**Comment by karussell** [comment]
> Just with removing functionality worries me when we may need it in the future. :smile: 

Ah, we break it and see what people report and then act as fast as possible to fix those use cases. No worries :)

There is one use case I could imagine that could be broken afterwards: using a protected variable inside the hook ... but we do not have `protected` vars in GraphHopper

**Comment by karussell** [comment]
Maybe instead of adding this bunch of code here we should try to fix #644 without breaking the old intended behaviour. Will have a look.

**Comment by karussell** [comment]
Found a way to fix #644 and keeping backward compatibility
